### PR TITLE
Add notes features

### PR DESCRIPTION
**Pull request by antoinelefebvre17**

### CODE REVIEW

**Review by zombre98 — CHANGES_REQUESTED** [review]
What's with all the .pyc files?
Can you remove it please

**Comment by antoinelefebvre17** [comment]
> What's with all the .pyc files?
> Can you remove it please

Done

File: app/features/MidiData.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+class MidiData:
+    def __init__(self, file_name):
+        self.mid = MidiFile(file_name)
+        self.tempos = Tempos()
+        self.notes = Notes()
+
+        self.parseData()
+
+    def parseData(self):
+        print("Data recovery...")
```
Comment (armandmgt) [inline app/features/MidiData.py]: Could we add a `verbose` option which defaults to `False` ?

File: app/features/MidiData.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+        self.notes.setTime(self.mid.length)
+        self.tempos.setTime(self.mid.length)
+        currentTime = 0
+        tempoTemp = 0
+        for msg in merge_tracks(self.mid.tracks):
+            currentTime = tick2second(msg.time, self.mid.ticks_per_beat, tempoTemp) + currentTime
```
Comment (armandmgt) [inline app/features/MidiData.py]: Minor comment, why don't you use the `+=` operator ?

File: app/features/MidiData.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+                find = False
+                for j in range(len(self.tempos.temposWithRepetition)):
+                    if self.tempos.temposWithRepetition[j].get('tempo') == round(tempo2bpm(msg.tempo)):
+                        self.tempos.temposWithRepetition[j]['repetition'] = self.tempos.temposWithRepetition[j][
+                                                                                  'repetition'] + 1
+                        self.tempos.temposWithRepetition[j]['time'] = tick2second(msg.time, self.mid.ticks_per_beat, msg.tempo) + self.tempos.temposWithRepetition[j]['time']
+                        find = True
+                if not find:
+                    self.tempos.temposWithRepetition.append(
+                        {'tempo': round(tempo2bpm(msg.tempo)), 'repetition': 1, 'time': tick2second(msg.time, self.mid.ticks_per_beat, msg.tempo)})
```
Comment (armandmgt) [inline app/features/MidiData.py]: Should this go inside the `Tempos` class ?
The class would directly have a method `append` that receives the tempo `msg` and` currentTime` and does this job of finding the repetitions, etc..

File: app/features/MidiData.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+                self.notes.notes.append({'note': msg.note, 'currentTime': currentTime})
+
+                find = False
+                for j in range(len(self.notes.notesWithRepetition)):
+                    if self.notes.notesWithRepetition[j].get('note') == msg.note:
+                        self.notes.notesWithRepetition[j]['repetition'] = self.notes.notesWithRepetition[j][
+                                                                                  'repetition'] + 1
+                        find = True
+                if not find:
+                    self.notes.notesWithRepetition.append({'note': msg.note, 'repetition': 1})
```
Comment (armandmgt) [inline app/features/MidiData.py]: Same for this part, maybe move the logic to the `Notes` class

File: app/features/Type/Notes.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
+class Notes:
+    def __init__(self):
+        self.notesWithRepetition = []  # {'tempo': , 'repetition': , 'time': }
+        self.notes = []  # {{'tempo': , 'currentTime': }
+        self.time = 0
+
+    def setTime(self, time):
```
Comment (armandmgt) [inline app/features/Type/Notes.py]: Maybe rename to `totalTime` ?

File: app/features/Type/Notes.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
+class Notes:
+    def __init__(self):
+        self.notesWithRepetition = []  # {'tempo': , 'repetition': , 'time': }
```
Comment (armandmgt) [inline app/features/Type/Notes.py]: Uh oh, `tempos` in the `Notes` class ?

File: app/features/MidiData.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+                if not find:
+                    self.tempos.temposWithRepetition.append(
+                        {'tempo': round(tempo2bpm(msg.tempo)), 'repetition': 1, 'time': tick2second(msg.time, self.mid.ticks_per_beat, msg.tempo)})
+            elif msg.type == 'note_on':
+                self.notes.notes.append({'note': msg.note, 'currentTime': currentTime})
+
+                find = False
+                for j in range(len(self.notes.notesWithRepetition)):
+                    if self.notes.notesWithRepetition[j].get('note') == msg.note:
+                        self.notes.notesWithRepetition[j]['repetition'] = self.notes.notesWithRepetition[j][
```
Comment (armandmgt) [inline app/features/MidiData.py]: `+= 1`

File: app/features/MidiData.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+            currentTime = tick2second(msg.time, self.mid.ticks_per_beat, tempoTemp) + currentTime
+            if msg.type == 'set_tempo':
+                tempoTemp = msg.tempo
+                # fill tempo
+                self.tempos.tempos.append({'tempo': round(tempo2bpm(msg.tempo)), 'currentTime': currentTime})
+                # fill tempo with repetition
+                find = False
+                for j in range(len(self.tempos.temposWithRepetition)):
+                    if self.tempos.temposWithRepetition[j].get('tempo') == round(tempo2bpm(msg.tempo)):
+                        self.tempos.temposWithRepetition[j]['repetition'] = self.tempos.temposWithRepetition[j][
```
Comment (armandmgt) [inline app/features/MidiData.py]: `+= 1`

File: app/features/MidiData.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+                                                                                  'repetition'] + 1
+                        self.tempos.temposWithRepetition[j]['time'] = tick2second(msg.time, self.mid.ticks_per_beat, msg.tempo) + self.tempos.temposWithRepetition[j]['time']
```
Comment (armandmgt) [inline app/features/MidiData.py]: `+= tick2second(...)`

**Review by armandmgt — APPROVED** [review]
👌

**Review by zombre98 — APPROVED** [review]

**Pull request merged**